### DISCUSSION TITLE
Report feedback: jerk-limited parameterization

### DISCUSSION THREAD

**Discussion by v4hn** [NVlabs/curobo#4] [Research Q&A] [answered]
Not sure this is a good place to give feedback about your report, but here it is. :dog: 

From your report page 16:

> Geometric planning methods require time parameterization as a post-processing step to be able to execute on robots.
Kunz and Stilman developed a bounded velocity and acceleration time-parameterization technique [4] that
is extensively used in the robotics community, including MoveIt [55]. However, this time parameterization
technique does not bound the jerk along the trajectory and as such can have very large jerks. We could not
find any accessible software library that can post process geometric paths while bounding jerks, making
geometric path planning not directly deploy-able on jerk sensitive manipulators.

The currently-advised way to limit jerks in MoveIt is to run a ruckig smoothing filter on the parameterized trajectory - at almost no overhead due to the performant implementation of ruckig. Here is the short tutorial you could not find (whether it's "accessible" definitely depends on your standards, I'm the last to claim MoveIt has good usability):
https://ros-planning.github.io/moveit_tutorials/doc/time_parameterization/time_parameterization_tutorial.html?highlight=ruckig#jerk-limited-trajectory-smoothing
I would (have) be(en) interested in the comparison.

**Comment by balakumar-s** [comment, marked as answer]
That text refers to Ruckig's Pro version that allows for time parameterization of trajectory with waypoints. 

Looking at MoveIt's ruckig integration ([code](https://github.com/ros-planning/moveit2/blob/63e0c3a3994c8eab5fff00daff00d4100df05383/moveit_core/trajectory_processing/src/ruckig_traj_smoothing.cpp#L273)), MoveIt doesn't  use the full trajectory feature from Ruckig Pro. It rather runs Ruckig from waypoint to waypoint with some branching to account for overshooting. We tried to do something similar with  Ruckig's python api but could not get good results.

I do agree that running the benchmarks with moveit would be good. We share the dataset with metrics computations here: https://github.com/fishbotics/robometrics

**Reply by v4hn** [reply]
I'm not sure where the Pro version is referred to, but I'd be happy to clean that up. :-)

Yes, MoveIt's ruckig smoothing filter only uses the open source version. I'm not convinced the results are good enough without thorough testing on relevant trajectories, but that's a different story.

Thanks for the link to robometrics. I've previously only worked with https://github.com/KavrakiLab/robowflex and https://github.com/captain-yoshi/moveit_benchmark_suite , both closer (or dedicated) to MoveIt.
